File: WalletKitJava/cryptodemo-android/src/main/java/com/breadwallet/cryptodemo/CoreSystemListener.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.breadwallet.cryptodemo;

import android.support.annotation.Nullable;

import com.breadwallet.crypto.Account;
import com.breadwallet.crypto.Address;
import com.breadwallet.crypto.AddressScheme;
import com.breadwallet.crypto.Coder;
import com.breadwallet.crypto.Currency;
import com.breadwallet.crypto.Network;
import com.breadwallet.crypto.NetworkType;
import com.breadwallet.crypto.System;
import com.breadwallet.crypto.Transfer;
import com.breadwallet.crypto.Wallet;
import com.breadwallet.crypto.WalletManager;
import com.breadwallet.crypto.WalletManagerMode;
import com.breadwallet.crypto.blockchaindb.models.bdb.HederaAccount;
import com.breadwallet.crypto.errors.AccountInitializationError;
import com.breadwallet.crypto.errors.AccountInitializationMultipleHederaAccountsError;
import com.breadwallet.crypto.events.network.NetworkEvent;
import com.breadwallet.crypto.events.system.DefaultSystemEventVisitor;
import com.breadwallet.crypto.events.system.SystemDiscoveredNetworksEvent;
import com.breadwallet.crypto.events.system.SystemEvent;
import com.breadwallet.crypto.events.system.SystemListener;
import com.breadwallet.crypto.events.system.SystemManagerAddedEvent;
import com.breadwallet.crypto.events.system.SystemNetworkAddedEvent;
import com.breadwallet.crypto.events.transfer.TranferEvent;
import com.breadwallet.crypto.events.wallet.DefaultWalletEventVisitor;
import com.breadwallet.crypto.events.wallet.WalletCreatedEvent;
import com.breadwallet.crypto.events.wallet.WalletEvent;
import com.breadwallet.crypto.events.walletmanager.WalletManagerEvent;
import com.breadwallet.crypto.utility.CompletionHandler;
import com.google.common.base.Optional;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import static com.google.common.base.Preconditions.checkState;

public class CoreSystemListener implements SystemListener {

    private static final Logger Log = Logger.getLogger(CoreSystemListener.class.getName());

    private final WalletManagerMode preferredMode;
    private final boolean isMainnet;
    private final List<String> currencyCodesNeeded;

    /* package */
    CoreSystemListener(WalletManagerMode preferredMode, boolean isMainnet, List<String> currencyCodesNeeded) {
        this.preferredMode = preferredMode;
        this.isMainnet = isMainnet;
        this.currencyCodesNeeded = new ArrayList<>(currencyCodesNeeded);
    }

    // SystemListener Handlers

    @Override
    public void handleSystemEvent(System system, SystemEvent event) {
        ApplicationExecutors.runOnBlockingExecutor(() -> {
            Log.log(Level.FINE, String.format("System: %s", event));

            event.accept(new DefaultSystemEventVisitor<Void>() {
                @Nullable
                @Override
                public Void visit(SystemNetworkAddedEvent event) {
                    createWalletManager(system, event.getNetwork());
                    return null;
                }

                @Nullable
                @Override
                public Void visit(SystemManagerAddedEvent event) {
                    connectWalletManager(event.getWalletManager());
                    return null;
                }

                @Nullable
                @Override
                public Void visit(SystemDiscoveredNetworksEvent event) {
                    logDiscoveredCurrencies(event.getNetworks());
                    return null;
                }
            });
        });
    }

    @Override
    public void handleNetworkEvent(System system, Network network, NetworkEvent event) {
        ApplicationExecutors.runOnBlockingExecutor(() -> {
            Log.log(Level.FINE, String.format("Network: %s", event));
        });
    }

    @Override
    public void handleManagerEvent(System system, WalletManager manager, WalletManagerEvent event) {
        ApplicationExecutors.runOnBlockingExecutor(() -> {
            Log.log(Level.FINE, String.format("Manager (%s): %s", manager.getName(), event));
        });
    }

    @Override
    public void handleWalletEvent(System system, WalletManager manager, Wallet wallet, WalletEvent event) {
        ApplicationExecutors.runOnBlockingExecutor(() -> {
            Log.log(Level.FINE, String.format("Wallet (%s:%s): %s", manager.getName(), wallet.getName(), event));

            event.accept(new DefaultWalletEventVisitor<Void>() {
                @Nullable
                @Override
                public Void visit(WalletCreatedEvent event) {
                    logWalletAddresses(wallet);
                    return null;
                }
            });
        });
    }

    @Override
    public void handleTransferEvent(System system, WalletManager manager, Wallet wallet, Transfer transfer, TranferEvent event) {
        ApplicationExecutors.runOnBlockingExecutor(() -> {
            Log.log(Level.FINE, String.format("Transfer (%s:%s): %s", manager.getName(), wallet.getName(), event));
        });
    }

    // Misc.

    private void createWalletManager(System system, Network network) {
        boolean isNetworkNeeded = false;
        for (String currencyCode : currencyCodesNeeded) {
            Optional<? extends Currency> currency = network.getCurrencyByCode(currencyCode);
            if (currency.isPresent()) {
                isNetworkNeeded = true;
                break;
            }
        }

        if (isMainnet == network.isMainnet() && isNetworkNeeded) {
            WalletManagerMode mode = network.supportsWalletManagerMode(preferredMode) ?
                    preferredMode : network.getDefaultWalletManagerMode();

            AddressScheme addressScheme = network.getDefaultAddressScheme();
            Log.log(Level.FINE, String.format("Creating %s WalletManager with %s and %s", network, mode, addressScheme));
            boolean success = system.createWalletManager(network, mode, addressScheme, Collections.emptySet());
            if (!success) {
                Account account = system.getAccount();

                system.wipe(network);
                if (!system.accountIsInitialized(account, network)) {

                    checkState (network.getType() == NetworkType.HBAR);
                    List<byte[]> serializationData = new ArrayList<>();

                    system.accountInitialize(system.getAccount(), network, true, new CompletionHandler<byte[], AccountInitializationError>() {
                        @Override
                        public void handleData(byte[] data) {
                            serializationData.add(data);
                            createWalletManagerIfAppropriate(serializationData, system, network, mode, addressScheme);
                        }

                        @Override
                        public void handleError(AccountInitializationError error) {
                            if (error instanceof AccountInitializationMultipleHederaAccountsError) {
                                List<HederaAccount> accounts = ((AccountInitializationMultipleHederaAccountsError) error).getAccounts();

                                // TODO: Sort accounts?

                                system.accountInitializeUsingHedera (system.getAccount(), network, accounts.get(0))
                                        .transform((bytes) -> { serializationData.add(bytes); return "ignore"; });
                            }
                            createWalletManagerIfAppropriate(serializationData, system, network, mode, addressScheme);
                        }
                    });

                }
            }
        }
    }

    private void createWalletManagerIfAppropriate (List<byte[]> serializationData,
                                                   System system,
                                                   Network network,
                                                   WalletManagerMode mode,
                                                   AddressScheme addressScheme) {
        if (!serializationData.isEmpty()) {
            Coder hexCoder = Coder.createForAlgorithm(com.breadwallet.crypto.Coder.Algorithm.HEX);

            // Normally, save the `serializationData`; but not here - DEMO-SPECIFIC
            Log.log(Level.INFO, String.format("Account: SerializationData: %s",
                    hexCoder.encode(serializationData.get(0))));

            checkState(system.createWalletManager(network, mode, addressScheme, Collections.emptySet()));
        }
    }

    private void connectWalletManager(WalletManager walletManager) {
        walletManager.connect(null);
    }

    private void logWalletAddresses(Wallet wallet) {
        Log.log(Level.FINE, String.format("Wallet (target) addresses: %s", wallet.getTarget()));
    }

    private void logDiscoveredCurrencies(List<Network> networks) {
        for (Network network: networks) {
            for (Currency currency: network.getCurrencies()) {
                Log.log(Level.FINE, String.format("Discovered: %s for %s", currency.getCode(), network.getName()));
            }
        };
    }
}

File: WalletKitJava/cryptodemo-android/src/main/java/com/breadwallet/cryptodemo/CoreSystemListener.kt
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
*
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.breadwallet.cryptodemo

import com.breadwallet.crypto.*
import com.breadwallet.crypto.errors.AccountInitializationError
import com.breadwallet.crypto.errors.AccountInitializationMultipleHederaAccountsError
import com.breadwallet.crypto.events.network.NetworkEvent
import com.breadwallet.crypto.events.system.*
import com.breadwallet.crypto.events.transfer.TranferEvent
import com.breadwallet.crypto.events.wallet.WalletCreatedEvent
import com.breadwallet.crypto.events.wallet.WalletEvent
import com.breadwallet.crypto.events.walletmanager.WalletManagerEvent
import com.breadwallet.crypto.utility.CompletionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class CoreSystemListener(
        private val preferredMode: WalletManagerMode,
        private val isMainnet: Boolean,
        private val currencyCodesNeeded: List<String>
) : SystemListener {
    companion object {
        private val Log = Logger.getLogger(CoreSystemListener::class.java.name)
    }

    private val scope = CoroutineScope(SupervisorJob() + IO)

    // SystemListener Handlers
    override fun handleSystemEvent(system: System, event: SystemEvent) {
        scope.launch {
            Log.log(Level.FINE, "System: $event")
            when (event) {
                is SystemNetworkAddedEvent ->
                    createWalletManager(system, event.network)
                is SystemManagerAddedEvent ->
                    connectWalletManager(event.walletManager)
                is SystemDiscoveredNetworksEvent ->
                    logDiscoveredCurrencies(event.networks)
            }
        }
    }

    override fun handleNetworkEvent(system: System, network: Network, event: NetworkEvent) {
        scope.launch {
            Log.log(Level.FINE, "Network: $event")
        }
    }

    override fun handleManagerEvent(system: System, manager: WalletManager, event: WalletManagerEvent) {
        scope.launch {
            Log.log(Level.FINE, "Manager (${manager.name}): $event")
        }
    }

    override fun handleWalletEvent(system: System, manager: WalletManager, wallet: Wallet, event: WalletEvent) {
        scope.launch {
            Log.log(Level.FINE, "Wallet (${manager.name}:${wallet.name}): $event")
            when (event) {
                is WalletCreatedEvent -> logWalletAddresses(wallet)
            }
        }
    }

    override fun handleTransferEvent(
            system: System,
            manager: WalletManager,
            wallet: Wallet,
            transfer: Transfer,
            event: TranferEvent?
    ) {
        scope.launch {
            Log.log(Level.FINE, "Transfer (${manager.name}:${wallet.name}): $event")
        }
    }

    // Misc.
    private suspend fun createWalletManager(system: System, network: Network) {
        val isNetworkNeeded = currencyCodesNeeded.any {
            network.getCurrencyByCode(it).isPresent
        }
        if (isMainnet == network.isMainnet && isNetworkNeeded) {
            val addressScheme = network.defaultAddressScheme
            val mode = when {
                network.supportsWalletManagerMode(preferredMode) -> preferredMode
                else -> network.defaultWalletManagerMode
            }
            Log.log(Level.FINE, "Creating $network WalletManager with $mode and $addressScheme")
            val success = system.createWalletManager(network, mode, addressScheme, emptySet())
            if (!success) {
                val account = system.account
                system.wipe(network)
                if (!system.accountIsInitialized(account, network)) {
                    check(network.type === NetworkType.HBAR)
                    try {
                        system.accountInitialize(account, network, true)
                    } catch (e: AccountInitializationError) {
                        (e as? AccountInitializationMultipleHederaAccountsError)?.run {
                            // TODO: Sort accounts?
                            system.accountInitializeUsingHedera(account, network, accounts.first())
                        }?.orNull()
                    }?.let { accountBytes ->
                        createWalletManager(accountBytes, system, network, mode, addressScheme)
                    }
                }
            }
        }
    }

    private fun createWalletManager(
            serializationData: ByteArray,
            system: System,
            network: Network,
            mode: WalletManagerMode,
            addressScheme: AddressScheme
    ) {
        if (serializationData.isNotEmpty()) {
            val hexCoder = Coder.createForAlgorithm(Coder.Algorithm.HEX)

            // Normally, save the `serializationData`; but not here - DEMO-SPECIFIC
            Log.log(Level.INFO, "Account: SerializationData: ${hexCoder.encode(serializationData)}")
            check(system.createWalletManager(network, mode, addressScheme, emptySet())) {
                "Failed to create wallet manager: $network, $mode, $addressScheme"
            }
        }
    }

    private fun connectWalletManager(walletManager: WalletManager) {
        walletManager.connect(null)
    }

    private fun logWalletAddresses(wallet: Wallet) {
        Log.log(Level.FINE, "Wallet (target) addresses: ${wallet.target}")
    }

    private fun logDiscoveredCurrencies(networks: List<Network>) {
        networks.forEach { network ->
            network.currencies.forEach { currency ->
                Log.log(Level.FINE, "Discovered: ${currency.code} for $network")
            }
        }
    }

    @Throws(AccountInitializationError::class)
    private suspend fun System.accountInitialize(
            account: Account,
            network: Network,
            create: Boolean
    ): ByteArray = suspendCoroutine { continuation ->
        accountInitialize(
                account,
                network,
                create,
                object : CompletionHandler<ByteArray, AccountInitializationError> {
                    override fun handleData(data: ByteArray) {
                        continuation.resume(data)
                    }

                    override fun handleError(error: AccountInitializationError) {
                        continuation.resumeWithException(error)
                    }
                })
    }
}